packages/muelu/src/Transfers/Smoothed-Aggregation/MueLu_SaPFactory_kokkos_def.hpp

```diff
@@ -271,53 +271,56 @@ namespace MueLu {
   // the constraints.
 
 template<typename local_matrix_type>
-class constraintKernel {
-public:
+struct constraintKernel {
+
    using Scalar= typename local_matrix_type::non_const_value_type;
    using SC=Scalar;
    using LO=typename local_matrix_type::non_const_ordinal_type;
-//   using Device= typename Matrix::local_matrix_type::device_type;
    using Device= typename local_matrix_type::device_type;
-   Scalar zero=Kokkos::ArithTraits<Scalar>::zero();
+   const Scalar zero = Kokkos::ArithTraits<SC>::zero();
+   const Scalar one  = Kokkos::ArithTraits<SC>::one();
    LO     nPDEs;
    local_matrix_type localP;
-   Kokkos::View<Scalar*, Device> ConstraintViolationSum;
-   Kokkos::View<Scalar*, Device> Rsum;
-   Kokkos::View<size_t*, Device> nPositive;
-
-   constraintKernel(LO nPDEs_,local_matrix_type localP_):nPDEs(nPDEs_), localP(localP_) {
-   ConstraintViolationSum = Kokkos::View<Scalar*, Device>("ConstraintViolationSum",nPDEs);
-   Rsum = Kokkos::View<Scalar*, Device>("Rsum",nPDEs);
-   nPositive = Kokkos::View<size_t*, Device>("nPositive",nPDEs);
+   Kokkos::View<SC**, Device> ConstraintViolationSum;
+   Kokkos::View<SC**, Device> Rsum;
+   Kokkos::View<size_t**, Device> nPositive;
+
+  constraintKernel(LO nPDEs_,local_matrix_type localP_) : nPDEs(nPDEs_), localP(localP_)
+   {
+     ConstraintViolationSum = Kokkos::View<SC**, Device>("ConstraintViolationSum", localP_.numRows(), nPDEs);
+     Rsum = Kokkos::View<SC**, Device>("Rsum", localP_.numRows(), nPDEs);
```

Comment:
Is this the ugly fix?

Reply:
Yep, I have to ask someone else on the kokkos team about thread local memory when one is not using a team policy...

Reply:
I don't think there's any way to use shared mem in a RangePolicy. Are you worried about lost performance from having ``ConstraintViolationSum`` and ``Rsum`` be in global memory instead? I would not expect it to make that much difference. You could also add ``Kokkos::LayoutRight`` to make the rows contiguous in memory, then caching will be better.